### DISCUSSION TITLE
$q.dialog  in Quasar2

### DISCUSSION THREAD

**Discussion by zanzara** [quasarframework/quasar#10209] [General - Components / Directives / etc]
The migration of my Q1 project into Q2 is driving me insane - really.
I got this little `logout` component below working in Q1 like charme and I grow desperate to get it work on Q2
Appreciating any further help.

```vue
<template>
  <q-page class="flex flex-center" />
</template>

<script>
export default {
  name: "Logout",
  mounted() {
    this.$q
      .dialog({
        title: this.$i18n.t("auth.logout.confirm"),
        message: this.$i18n.t("auth.logout.logout_confirmation"),
        ok: this.$i18n.t("auth.logout.logout"),
        cancel: this.$i18n.t("auth.logout.cancel")
      })
      .onOk(() => {
        this.$auth.logout().then(() => {
          this.$router.push({name: 'home'});
        });
      })
      .onCancel(() => {
        this.$router.go(-1);
      });
  }
};
</script>

``` 
What is the proper way to migrate this?

**Comment by stefanvanherwijnen** [comment]
Did you add `Dialog` to your quasar.conf.js plugins?
Any errors would be helpful to figure out what is wrong.

**Reply by zanzara** [reply]
Yes, of course. 
Unfortunately no error msg appears in the console.  Only nothing happens. Blank screen.

But meanwhile I made some more investigations and have been little successful. 

```vue
<template>
  <q-page class="flex flex-center" />
</template>
<script>
import { useQuasar } from "quasar";
import { useI18n } from 'vue-i18n'
import { useRouter } from "vue-router";
import {  inject } from 'vue'

export default {
  name: "Logout",
  setup() {
    const $q = useQuasar()
    const { t } = useI18n({ useScope: 'global' })
    const $auth = inject('auth');
    const $router = useRouter()

    $q.dialog({
      title: t("auth.logout.confirm"),
      message: t("auth.logout.logout_confirmation"),
      ok: t("auth.logout.logout"),
      cancel: t("auth.logout.cancel")
    }).onOk(() => {
        $auth.logout().then(() => {
          $router.push({name: 'home'});
        });
      }).onCancel(() => {
        $router.go(-1);
      });
  }
};
</script>

```
So , now it seems to be that I got.
But imo it's more cumbersome than with Q1(Vue2)

**Reply by stefanvanherwijnen** [reply]
This looks correct. The composition API  is less consise but much easier to use and debug imo. There is no `this`, but the contents of it are defined in a differemt fille. So with your original example it is impossible to know if `this` contains the necessary objects.

I'd definitely recommemd the composition API, but it's a different approach from the Options API.

**Reply by zanzara** [reply]
You maybe right. But I got until now too much breaking changes in my project. Incredible.
The next issue I got now is struggling with `vuelidate` :-(

(you could mark this answered, if you want.  I don't know if I can do)

**Reply by stefanvanherwijnen** [reply]
I just noticed an error in your original example. I think the correct syntax is `this.$q.create(...)`, instead of `this.$q`.

For vuelidate: https://vuelidate-next.netlify.app/. Make sure you use vuelidate v2.

I would also recommend writing something from scratch if you want to switch to the composition API. If you are trying to migrate a v1 project to v2 you will indeed get a lot of errors if you simply try to copy things (as Vue 3 introduced breaking changes for most dependencies). I've tried it, doesn't work well :grin: .

**Reply by zanzara** [reply]
At least my original example works well in production. Imo it's either using `Dialog.create()` or `this.$q.dialog()`  .

My Q1(Vue2) project is too big to write it complete from the scratch to consider all Q2(Vue3) purposes and changes.
I followed the guide and at least I got the main page running. But I'm curious when it comes to my `capacitor` artefact.

Regarding `vuelidate` it works in general but i got issues now with `sameAs` builtin.  Cause it does not accept an empty, none-required `sameAs` property .  Now I'm trying to fix it with `requiredIf` but it does not work

**Reply by stefanvanherwijnen** [reply]
> At least my original example works well in production. Imo it's either using `Dialog.create()` or `this.$q.dialog()` .

You are right, my bad...

The problem is you cannot ignore most of the v2 (Vue 3) changes or it will break your project, as you have noticed. I'd advice getting familiar with v2/Vue 3 first before migrating a large project.

Vue 3 also in general seems more strict about for example not accepting non-defined values in functions. This is actually the correct behavior, but I also struggled with the switch.
E.g. `sameAs(nonDefined)` should be `nonDefined ? sameAs(nonDefined) : false` (check if the value is defined before calling the function). When working with TypeScript it will be clear why it works like this, because if a value could be undefined and you don't make sure that it is defined, you will receive an error.
```
const maybeUndefined: number | undefined = getValueFromSomewhereElse()
sameAs(maybeUndefined) // will report an error: maybeUndefined might be undefined
```

**Reply by zanzara** [reply]
Yes, starting something from the scratch would be more helpful to understand all the breaking changes in Vue3.
Fortunately, most dependend modules supports both: composition API and option API.

Considering `vuelidate` .  These are my validations:

```javascript
  validations: {
    form: {
      oldpassword: {
        required
      },
      newpassword: {
        maxLength: maxLength(userConst().maxPasswordLength),
        minLength: minLength(userConst().minPasswordLength)
      },
      repeatPassword: {
        sameAsPassword: sameAs('newpassword')
      },
      prename: {
        required,
        maxLength: maxLength(userConst().maxNameLength)
      },
      surname: {
        required,
        maxLength: maxLength(userConst().maxNameLength)
      },
      email: {
        email,
        required,
        maxLength: maxLength(userConst().maxEmailLength)
      }
    }
  }
```
 If I left `newpassword` and `repeatPassword` empty, I got an `this.v$.form.$error` which says, that they are not the same
(I didn't got this behaviour in V1.)

Sry, for getting OT. Hoping, I'm not getting railed now.

**Reply by stefanvanherwijnen** [reply]
I don't use vuelidate but I think same As does not work like you think it does : https://vuelidate-next.netlify.app/validators.html#sameas

In your example you are comparing to the string 'newpassword' according to the docs.

**Reply by zanzara** [reply]
Yes, there must be something different to V1.
Pls look here [https://vuelidate.js.org/#sub-form-submission](https://vuelidate.js.org/#sub-form-submission)  under `Contextified validators` .

**Reply by zanzara** [reply]
@stefanvanherwijnen   To cut this story finally, this way works now:
```javascript
 validations() {
    return {
      form: {
        oldpassword: {
          required
        },
        newpassword: {
          maxLength: maxLength(userConst().maxPasswordLength),
          minLength: minLength(userConst().minPasswordLength)
        },
        repeatPassword: {
          sameAsPassword: sameAs(this.form.newpassword),
        },
        prename: {
          required,
          maxLength: maxLength(userConst().maxNameLength)
        },
        surname: {
          required,
          maxLength: maxLength(userConst().maxNameLength)
        },
        email: {
          email,
          required,
          maxLength: maxLength(userConst().maxEmailLength)
        }
      }
    }
  }
```
the issue was, that in V2 `validations`  has to be a function, not a property. Then one can use `this....`  within `sameAs`
